Google.PowerShell/Common/ActiveUserCredential.cs

```diff
@@ -0,0 +1,172 @@
+﻿/*
+Copyright 2016 Google Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Http;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.PowerShell.Common
+{
+    /// <summary>
+    /// OAuth 2.0 credential for accessing protected resources using an access token.
+    /// This class will get the access token from "gcloud auth print-access-token".
+    /// </summary>
+    public class ActiveUserCredential : ICredential, IHttpExecuteInterceptor, IHttpUnsuccessfulResponseHandler
+    {
+        private static ActiveUserToken token;
+        private static object lockObject = new object();
+
+        /// <summary>
+        /// We just need a default flow for things like flow.AccessMethod.Intercept
+        /// or flow.RevokeTokenAsync.
+        /// </summary>
+        private readonly IAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow(
+                new GoogleAuthorizationCodeFlow.Initializer()
+                {
+                    ClientSecrets = new ClientSecrets()
+                    {
+                        ClientId = "clientId",
+                        ClientSecret = "clientSecrets"
+                    }
+                });
+
+        /// <summary>UserId is not important, it can be anything.</summary>
+        private readonly string userId = "userId";
+
+        /// <summary>Gets or sets the token response which contains the access token.</summary>
+        private static ActiveUserToken Token
+        {
+            get
+            {
+                lock (lockObject)
+                {
+                    return token;
+                }
+            }
+            set
+            {
+                lock (lockObject)
+                {
+                    token = value;
+                }
+            }
+        }
+
+        /// <summary>Gets the authorization code flow.</summary>
+        public IAuthorizationCodeFlow Flow
+        {
+            get { return flow; }
+        }
+
+        /// <summary>Constructs a new credential instance.</summary>
+        public ActiveUserCredential()
+        {
+            if (Token == null)
+            {
+                Token = GCloudWrapper.GetAccessToken().Result;
```

Comment:
The problem with this is that it blocks. Perhaps we should just leave this as a TODO for now, but I'd imagine having `Token` only provide a getter, and it simply returns `tokenRequestTask`. (`Task<ActiveUserToken> tokenRequestTask`.)

When you access Token, it would call `tokenRequestTask.Result`, meaning that we don't block in the construction of `ActiveUserCredential` until the credential is returned.

Just a thought.